src/viur/core/modules/file.py

```diff
@@ -766,104 +799,121 @@ def download(self, blobKey: str, fileName: str = "", download: str = "", sig: st
+                response.headers["Content-Type"] = blob.content_type
+                response.headers["Cache-Control"] = "public, max-age=604800"  # 7 Days
+                if content_disposition:
+                    response.headers["Content-Disposition"] = content_disposition
+                return blob.download_as_bytes()
+
+        # Default fallback - create a signed URL and redirect
+        authRequest = google.auth.transport.requests.Request()
+        expiresAt = datetime.datetime.now() + datetime.timedelta(seconds=60)
+        signing_credentials = google.auth.compute_engine.IDTokenCredentials(authRequest, "")
+        signedUrl = blob.generate_signed_url(
+            expiresAt,
+            credentials=signing_credentials,
+            response_disposition=content_disposition,
+            version="v4")
+
+        raise errors.Redirect(signedUrl)
 
     @exposed
     @force_ssl
     @force_post
     @skey(allow_empty=True)
     def add(self, skelType: SkelType, node: db.Key | int | str | None = None, *args, **kwargs):
-        ## We can't add files directly (they need to be uploaded
-        # if skelType != "node":
-        #    raise errors.NotAcceptable()
+        # We can't add files directly (they need to be uploaded
         if skelType == "leaf":  # We need to handle leafs separately here
             targetKey = kwargs.get("key")
             skel = self.addSkel("leaf")
+
             if not skel.fromDB(targetKey):
                 raise errors.NotFound()
+
             if not skel["pending"]:
                 raise errors.PreconditionFailed()
+
             skel["pending"] = False
             skel["parententry"] = skel["pendingparententry"]
+
             if skel["parententry"]:
                 rootNode = self.getRootNode(skel["parententry"])
             else:
                 rootNode = None
+
             if not self.canAdd("leaf", rootNode):
                 # Check for a marker in this session (created if using a signed upload URL)
                 session = current.session.get()
                 if targetKey not in (session.get("pendingFileUploadKeys") or []):
                     raise errors.Forbidden()
                 session["pendingFileUploadKeys"].remove(targetKey)
                 session.markChanged()
-            blobs = list(bucket.list_blobs(prefix=f"""{skel["dlkey"]}/"""))
+
+            blobs = list(GOOGLE_STORAGE_BUCKET.list_blobs(prefix=f"""{skel["dlkey"]}/"""))
             if len(blobs) != 1:
                 logging.error("Invalid number of blobs in folder")
                 logging.error(targetKey)
                 raise errors.PreconditionFailed()
+
             blob = blobs[0]
             skel["mimetype"] = utils.string.escape(blob.content_type)
-            if any([x in blob.name for x in "$<>'\""]):  # Prevent these Characters from being used in a fileName
+
+            if any([ch in blob.name for ch in "$<>'\""]):  # Prevent these Characters from being used in a fileName
```

Comment:
Shouldn't this be the new method?

```suggestion
            if not self.is_valid_filename(blob.name):
```

Reply:
I wonder why this condition is here, as the filename of the blob checked when the upload URL is being provided. I think we can remove this entire thing? There's also no helpful comment about *why* this is being done here.

Reply:
with `getUploadURL` you only request a url for upload and say that the file has certain metadata. Which data you finally send to this upload URL is then independent. Therefore, everything (size, filename, mimetype) must be checked again at this point.

Reply:
But the check cloud also be
```py
if blob.name != skel["name"].removesuffix(PENDING_POSTFIX):
    raise ValueError("this is not the file you asked for in getUploadUrl")

Reply:
Hmm okay... then let's do it!